src/DI/ContainerBuilder.php

```diff
@@ -463,9 +468,14 @@ private function resolveEntityClass($entity, $recursive = [])
 				throw new ServiceCreationException(sprintf("Factory '%s' used in service '%s' is not callable.", Nette\Utils\Callback::toString($entity), $name[0]));
 			}
 
-			$class = preg_replace('#[|\s].*#', '', PhpReflection::parseAnnotation($reflection, 'return'));
-			if ($class) {
-				$class = $refClass ? PhpReflection::expandClassName($class, $refClass) : ltrim($class, '\\');
+			if (PHP_VERSION_ID >= 70000 && $reflection->hasReturnType()) {
+				$type = $reflection->getReturnType();
+				$class = !$type->isBuiltIn() ? (string) $type : null;
```

Comment:
Hehe, why can't I come up with these simple solutions. Thanks for the idea! It will be probably better than the solution I pushed, so I'll revert it and change `expandClassName`.